core/state_processor.go

```diff
@@ -1014,6 +1017,25 @@ func ProcessQiTx(tx *types.Transaction, chain ChainContext, updateState bool, ch
 		etxs = append(etxs, &etxInner)
 		txFeeInQit.Sub(txFeeInQit, txFeeInQit) // Fee goes entirely to gas to pay for conversion
 	}
+	// Go through all denominations largest to smallest, check if the input exists as the output, if not, convert it to the respective number of bills for the next smallest denomination, then repeat the check. Subtract the 'carry' when the outputs match the carry for that denomination.
```

Comment:
This logic is incorrect. You should use:

	carries := make(map[uint]uint64)
	for i := types.MaxDenomination; i >= 0; i-- {
		if outputs[uint(i)] <= inputs[uint(i)] + carries[uint(i)] {
			diff := new(big.Int).SetUint64((inputs[uint(i)] + carries[uint(i)]) - (outputs[uint(i)]))
			carries[uint(i-1)] += diff.Mul(diff, (new(big.Int).Div(types.Denominations[uint8(i)], types.Denominations[uint8(i-1)]))).Uint64()
		} else {
			return nil, nil, fmt.Errorf("tx attempts to combine smaller denominations into larger one for denomination %d", i)
		}
	}

Reply:
Also, you should put this into a function. We will probably also need to expose it.